overlord/devicestate/users.go

```diff
@@ -0,0 +1,372 @@
+	"github.com/snapcore/snapd/strutil"
+)
+
+type CreatedUser struct {
+	Username string
+	SSHKeys  []string
+}
+
+var (
+	osutilAddUser = osutil.AddUser
+	osutilDelUser = osutil.DelUser
+	userLookup    = user.Lookup
+)
+
+type UserError struct {
+	Err      error
+	Internal bool
+}
+
+func (e *UserError) Error() string {
+	return e.Err.Error()
+}
+
+func (e UserError) IsInternal() bool {
+	return e.Internal
+}
+
+// RemoveUser removes linux user account of passed username.
+func RemoveUser(st *state.State, username string) (*auth.UserState, *UserError) {
+	// TODO: allow to remove user entries by email as well
+
+	// catch silly errors
+	if username == "" {
+		return nil, &UserError{Internal: false, Err: fmt.Errorf("need a username to remove")}
+	}
+	// check the user is known to snapd
+	st.Lock()
+	_, err := auth.UserByUsername(st, username)
+	st.Unlock()
+	if err == auth.ErrInvalidUser {
+		return nil, &UserError{Internal: false, Err: fmt.Errorf("user %q is not known", username)}
+	}
+	if err != nil {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+
+	// first remove the system user
+	if err := osutilDelUser(username, &osutil.DelUserOptions{ExtraUsers: !release.OnClassic}); err != nil {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+
+	// then the UserState
+	st.Lock()
+	u, err := auth.RemoveUserByUsername(st, username)
+	st.Unlock()
+	// ErrInvalidUser means "not found" in this case
+	if err != nil && err != auth.ErrInvalidUser {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+	return u, nil
+}
+
+// CreateKnownUsers create known user(s)
+// user details are fetched from existing system user assertions.
+// If no email is passed, all known users will be created based on valid system user assertions.
+// If email is passed, only corresponding system user assertion is used for user creation.
+func CreateKnownUsers(st *state.State, mgr *DeviceManager, sudoer bool, email string) ([]CreatedUser, *UserError) {
+	var err error
+	st.Lock()
+	model, err := mgr.Model()
+	st.Unlock()
+	if err != nil {
+		return nil, &UserError{Internal: true, Err: fmt.Errorf("cannot create user: cannot get model assertion: %v", err)}
+	}
+	st.Lock()
+	serial, err := mgr.Serial()
+	st.Unlock()
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, &UserError{Internal: true, Err: fmt.Errorf("cannot create user: cannot get serial: %v", err)}
+	}
+
+	st.Lock()
+	db := assertstate.DB(st)
+	st.Unlock()
+
+	u := &createUserOpts{
+		assertDb: db,
+		modelAs:  model,
+		serialAs: serial,
+		isSudoer: sudoer,
+	}
+
+	if email == "" {
+		return u.createAllKnownSystemUsers(st)
+	}
+
+	username, opts, err := getUserDetailsFromAssertion(u.assertDb, u.modelAs, u.serialAs, email)
+
+	if err != nil {
+		return nil, &UserError{Internal: false, Err: err}
+	}
+	opts.Sudoer = sudoer
+	createdUser, userErr := u.createUser(st, opts, username, email)
+	var createdUsers []CreatedUser
+	createdUsers = append(createdUsers, createdUser)
```

Comment:
this can be done in one go with []CreatedUser{...}

Reply:
Fixed this, nice suggestion!